cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -332,9 +336,8 @@ private void processMetrics() {
         try {
           Deque<MetricValues> recordsCopy = new LinkedList<>();
           Map<TopicIdMetaKey, byte[]> topicMessageIdsCopy = new HashMap<>(topicMessageIds);
-          // TODO: (CDAP-8327) there is a risk of running out-of-memory if other threads keep writing to records
           Iterator<MetricValues> iterator = records.iterator();
-          while (iterator.hasNext()) {
+          while (iterator.hasNext() && recordsCopy.size() < persistThreshold) {
```

Comment:
I think a better logic would be this:

1. Turn the shared queue into a blocking queue. The size is configurable.
2. Each thread will fetch up to (blocking queue size / number of threads) messages from TMS. Use the `put` method to enqueue to block if the queue is full.
3. The leader writer thread will only deque up to the blocking queue size (using the `poll` method).
4. The sleep should be calculated based on how much time was spent in fetching and writing (for the leader).